Add proper packaging using rollup to all plugins, core and lib packages

dopenguin opened the pull request:
## Summary

- All plugin, core and lib packages consumed are now bundled with `rollup` to be able to use each package outside of this repository (I will add example repositories once this PR is merged)
- If a package defines some css rules, they are now bundled in a separate file and have to be imported where the package is consumed to pull them into the shadowDOM
- `lerna` has been removed in favour of `npm workspaces`
- `nx` is used for task scheduling
- Except for the client code, all type errors have been resolved; do you think running a `tsc --noEmit` task in the pipeline would be a good idea, after we fixed the type issues in the client code as well, @warm-coolguy?

## Instructions for local reproduction and review

**All** clients and **every** functionality are affected by the changes. Thus, all clients should be tested in `dev` as well as in `prod` (build) mode.  
If you got any questions while testing, feel free to ask me any time (I understand that the changes are quite *extensive*)...

## TODO

- [x] Add changelog entries for all 24 packages

## Relevant tickets, issues, et cetera

#107 has to be merged before this PR

warm-coolguy commented:
>Except for the client code, all type errors have been resolved; do you think running a tsc --noEmit task in the pipeline would be a good idea, after we fixed the type issues in the client code as well, @warm-coolguy?

Yes. It'd nice to keep that progress intact. Is this a rhetoric question or do you see any downside to it?

warm-coolguy reviewed (changes requested):
I've done a code-level review with minimal testing. A full functionality check will follow after we've sorted the comments so far out, lest we run into requiring a second full check.

dopenguin commented on package.json:
```diff
@@ -74,22 +88,29 @@
     "vue-jest": "^3.0.7",
     "vue-template-compiler": "^2.7.16"
-  }
+  },
+  "dependencies": {}
```
It was automatically added by nx

dopenguin commented on package.json:
```diff
@@ -74,22 +88,29 @@
+    "rollup-plugin-auto-external": "^2.0.0",
+    "rollup-plugin-scss": "^4.0.0",
+    "rollup-plugin-svg": "^2.0.0",
+    "rollup-plugin-typescript2": "^0.36.0",
```
I was testing multiple typescript plugins for rollup when I finally landed on this one, which worked.

I can test the other ones again, if you see the need.

warm-coolguy replied:
I'd prefer the "official" plugin, if you can make it work? Should this prove to be an issue/consume a lot of time, we can also roll with this underground one, but its declining popularity may bite later.

🎩 in case it "just doesn't work atm".

dopenguin replied:
I now remember why `@rollup/plugin-typescript` can not be used.  
If I read the documentation correctly, there is no option to not check with the rollup plugin for type errors. This leads to the problem, that it also want's to convert the vue files to js, which utterly fails.

It might be possible to change this behaviour once we upgrade to vue@3, as we'd just be using vite in that case, but for now I suggest that we keep using the unofficial plugin.

dopenguin commented on packages/core/rollup.config.js:
I'd suggest doing that in a separate iteration - after the packages have been tested outside of this repository.  
If you agree, feel free to create an issue and assign it to me

warm-coolguy replied:
https://github.com/Dataport/polar/issues/111

I created it with reference to this thread. Since it isn't urgent, I didn't assign it for the time being. 🔝👒

dopenguin replied:
Thank you!

dopenguin commented:
> > Except for the client code, all type errors have been resolved; do you think running a tsc --noEmit task in the pipeline would be a good idea, after we fixed the type issues in the client code as well, @warm-coolguy?
> 
> Yes. It'd nice to keep that progress intact. Is this a rhetoric question or do you see any downside to it?

Rather rhetorical, added so that we don't forget tha we need to do that.

warm-coolguy commented on packages/lib/getCluster/package.json:
```diff
@@ -1,16 +1,36 @@
 {
   "name": "@polar/lib-get-cluster",
   "version": "1.0.0",
+  "description": "Helper function to retrieve the related cluster of a feature.",
```
https://github.com/Dataport/polar/issues/112 🎩

warm-coolguy commented:
> > > Except for the client code, all type errors have been resolved; do you think running a tsc --noEmit task in the pipeline would be a good idea, after we fixed the type issues in the client code as well, @warm-coolguy?
> > 
> > 
> > Yes. It'd nice to keep that progress intact. Is this a rhetoric question or do you see any downside to it?
> 
> Rather rhetorical, added so that we don't forget tha we need to do that.

I know just the place for this stuff.

warm-coolguy commented:
> 🏓 @warm-coolguy

🏓 @dopenguin

dopenguin commented:
> > 🏓 @warm-coolguy
> 
> 🏓 @dopenguin

🏓 @warm-coolguy

warm-coolguy reviewed (changes requested):
```sh
vite v5.1.3 building for production...
✓ 45 modules transformed.
x Build failed in 738ms
error during build:
Error: [commonjs--resolver] Missing "./src/utils/searchResultSymbols" specifier in "@polar/plugin-address-search" package
    at e (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:25)
    at n (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:627)
    at o (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:1297)
    at resolveExportsOrImports (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:48032:20)
    at resolveDeepImport (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:48051:31)
    at tryNodeResolve (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47776:20)
    at Object.resolveId (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47539:28)
    at file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:18670:40
    at async PluginDriver.hookFirstAndGetPlugin (file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:18570:28)
    at async resolveId (file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:17239:26)
npm ERR! Lifecycle script `build` failed with error:
npm ERR! Error: command failed
npm ERR!   in workspace: @polar/client-dish@1.1.1
npm ERR!   at location: C:\Repos\polar-beeeg-review\packages\clients\dish

 ———————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————— 

 >  NX   Ran target build for project @polar/client-dish and 19 task(s) they depend on (4s)

    ×    1/20 failed
    √    19/20 succeeded [19 read from cache]
```

`npm run dish:build` (and, with an error alike, `npm run dish:dev`) fail with the above. I've compared it to `main`, where it's fine. Going `npm run clean && npm i && npm run dish:build` did not suffice to resolve the error, so it's not a hiccup.

🏓 @dopenguin

dopenguin commented:
> ```shell
> vite v5.1.3 building for production...
> ✓ 45 modules transformed.
> x Build failed in 738ms
> error during build:
> Error: [commonjs--resolver] Missing "./src/utils/searchResultSymbols" specifier in "@polar/plugin-address-search" package
>     at e (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:25)
>     at n (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:627)
>     at o (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47355:1297)
>     at resolveExportsOrImports (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:48032:20)
>     at resolveDeepImport (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:48051:31)
>     at tryNodeResolve (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47776:20)
>     at Object.resolveId (file:///C:/Repos/polar-beeeg-review/node_modules/vite/dist/node/chunks/dep-stQc5rCc.js:47539:28)
>     at file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:18670:40
>     at async PluginDriver.hookFirstAndGetPlugin (file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:18570:28)
>     at async resolveId (file:///C:/Repos/polar-beeeg-review/node_modules/rollup/dist/es/shared/node-entry.js:17239:26)
> npm ERR! Lifecycle script `build` failed with error:
> npm ERR! Error: command failed
> npm ERR!   in workspace: @polar/client-dish@1.1.1
> npm ERR!   at location: C:\Repos\polar-beeeg-review\packages\clients\dish
> 
>  ———————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————— 
> 
>  >  NX   Ran target build for project @polar/client-dish and 19 task(s) they depend on (4s)
> 
>     ×    1/20 failed
>     √    19/20 succeeded [19 read from cache]
> ```
> 
> `npm run dish:build` (and, with an error alike, `npm run dish:dev`) fail with the above. I've compared it to `main`, where it's fine. Going `npm run clean && npm i && npm run dish:build` did not suffice to resolve the error, so it's not a hiccup.
> 
> This does not happen in Meldemichel/Snowbox, despite them using `@polar/plugin-address-search`, too.
> 
> > > > 🏓 @warm-coolguy
> > > 
> > > 
> > > 🏓 @dopenguin
> > 
> > 
> > 🏓 @warm-coolguy
> 
> 🏓 @dopenguin
> 
> Also, there's an open TODO up there.

The problem only occurs in DISH as that is the only client that steps into the package to retrieve a different import.

There are two possible solutions for this now:
- Expose `SearchResultSymbols` through `index.ts` to have a named import like `import { searchResultSymbols } from '@polar/plugin-address-search'`
- Define `SearchResultSymbols` as a separate import like `import SearchResultSymbols from '@polar/plugin-address-search/searchResultSymbols'`

I prefer the second option, as the first option needs an export like

```ts
import exportResultSymbols from './utils/exportResultSymbols'

export { exportResultSymbols }
```

whereas the second option only changes the underlying setup and how the package is consumed - not the package itself.

Which option do you prefer?

🏓 @warm-coolguy

dopenguin commented:
> * import { searchResultSymbols } from '@polar/plugin-address-search'

As discussed on ☎️, the first option is preferred.  
I also added a proposed [ADR](https://github.com/Dataport/polar/wiki/Architecture-Decision-Records-%F0%9F%93%83#2024-03-05-how-to-expose-additional-exports-of-a-package).

If you don't see any additional issues, I will release a few alpha versions to test the functionality in another repository.

🏓 @warm-coolguy

warm-coolguy commented:
> As discussed on ☎️, the first option is preferred. I also added a proposed [ADR](https://github.com/Dataport/polar/wiki/Architecture-Decision-Records-%F0%9F%93%83#2024-03-05-how-to-expose-additional-exports-of-a-package).
> 
> If you don't see any additional issues, I will release a few alpha versions to test the functionality in another repository.
> 
> 🏓 @warm-coolguy

I've set the ADR to "Accepted.".

I find the order of entries in the Changelogs to be confusing. Until now, we've ordered them by impact, "Breaking > Feature > Patch > Chore". Breaking only occured once in PascalCase in `@polar/lib-components`, which I'd suggest to keep using as topmost list entry.

This minor detail aside, releasing alpha versions sounds great.

dopenguin commented:
To be able to yield a benefit from the changes introduced in this PR, either `vuetify` would have to be removed or an upgrade to `vue@3` *may* lead to usable packages.
As this is not an effort currently on the roadmap, this PR is closed for now and may be used as a reference point in the future.
